configure and build Positron

mykmelez opened the pull request:
@brendandahl This configures and (almost) builds Positron with Node automatically. On this branch, and with the changes in https://github.com/brendandahl/spidernode/pull/1, you can build with:

``` bash
export CXXFLAGS="-mmacosx-version-min=10.9"
export LDFLAGS="-mmacosx-version-min=10.9"
./mach build
make -C obj.debug.noindex/config/external/spidernode/
./mach build
```

The next bit to figure out is how to get config/external/spidernode/ to build automatically.

mykmelez committed:
configure and (theoretically) build spidernode

mykmelez committed:
integrate SpiderNode dependency into positron/webidl/ build config

mykmelez commented:
f1779fb makes SpiderNode build automatically, so the build instructions are now:

``` bash
export CXXFLAGS="-mmacosx-version-min=10.9"
export LDFLAGS="-mmacosx-version-min=10.9"
./mach build
```

This doesn't support release builds yet, nor have I tested on Linux. But I think that's true for the upstream branch as well. ;-)

I'm also unsure that this is the "right" way to do this. Some other third-party libraries are built via different techniques, and a build peer may be able to suggest a better approach. But it does configure and build SpiderNode when configuring and building Positron, so it's a good start.

mykmelez committed:
make flags additive

mykmelez committed:
set -mmacosx-version-min=10.9 only on Mac

mykmelez committed:
subconfigure SpiderNode via AC_OUTPUT_SUBDIRS call

mykmelez committed:
build SpiderNode release configuration if MOZ_DEBUG not defined

mykmelez commented on layout/base/nsDisplayList.cpp:
```diff
@@ -109,7 +109,7 @@ SpammyLayoutWarningsEnabled()
 
 void*
-AnimatedGeometryRoot::operator new(size_t aSize, nsDisplayListBuilder* aBuilder)
+AnimatedGeometryRoot::operator new(size_t aSize, nsDisplayListBuilder* aBuilder) CPP_THROW_NEW
```
Note: this is an unrelated change for a bug that I happened to run into while developing this patch. I'm unsure why I ran into it, but in any case I've fixed it upstream in https://bugzilla.mozilla.org/show_bug.cgi?id=1291889, and it shouldn't hurt to merge it here in the meantime.

mykmelez committed:
copy lib files to objdir instead of linking .libs to srcdir

mykmelez commented:
@brendandahl With the latest changes, this now builds both debug and release builds, and it copies libs to the objdir instead of linking _.libs_ back to the srcdir.

There's undoubtedly more to do (like adding a "clean::" recipe). And I haven't tested on Linux yet. But in theory, this is minimally sufficient to configure, build, and link SpiderNode with Positron, provided that you've cloned it to the right place. So I've removed the "WIP" tag from its summary. Let me know if there's anything else you think we should do here before merging it!

brendandahl commented:
After a clobber and a distclean in spidernode and then ./mach build I get:

```
 2:21.63 make: *** /Users/bdahl/projects/positron/positron/spidernode/out: No such file or directory.  Stop.
 2:21.63 make[5]: *** [spidernode] Error 2
 2:21.63 make[4]: *** [positron/webidl/target] Error 2
 2:21.63 make[4]: *** Waiting for unfinished jobs....
```

mykmelez commented:
> After a clobber and a distclean in spidernode and then ./mach build I get:

Hmm, I'm not seeing this. I'll try recloning from scratch to see if I can reproduce.

mykmelez commented:
Hmm, I still can't reproduce after cloning from scratch. Here's what I did:

``` bash
> git clone git@github.com:mykmelez/positron.git
Cloning into 'positron'...
remote: Counting objects: 5184985, done.
remote: Compressing objects: 100% (52/52), done.
remote: Total 5184985 (delta 28), reused 3 (delta 3), pack-reused 5184929
Receiving objects: 100% (5184985/5184985), 2.21 GiB | 441.00 KiB/s, done.
Resolving deltas: 100% (4179310/4179310), done.
Checking connectivity... done.
Checking out files: 100% (142342/142342), done.
> cd positron
> git checkout wip-apply-spidernode-patches3
Checking out files: 100% (7144/7144), done.
Branch wip-apply-spidernode-patches3 set up to track remote branch wip-apply-spidernode-patches3 from origin.
Switched to a new branch 'wip-apply-spidernode-patches3'
> git clone https://github.com/brendandahl/spidernode.git positron/spidernode
Cloning into 'positron/spidernode'...
remote: Counting objects: 248053, done.
remote: Compressing objects: 100% (42/42), done.
remote: Total 248053 (delta 17), reused 0 (delta 0), pack-reused 248011
Receiving objects: 100% (248053/248053), 231.19 MiB | 2.20 MiB/s, done.
Resolving deltas: 100% (173138/173138), done.
Checking connectivity... done.
Checking out files: 100% (44216/44216), done.
> cd positron/spidernode
> git checkout wip-link2
Checking out files: 100% (31683/31683), done.
Branch wip-link2 set up to track remote branch wip-link2 from origin.
Switched to a new branch 'wip-link2'
> cd ../..
> edit .mozconfig
#
# Added:
# ac_add_options --disable-optimize
# ac_add_options --enable-debug
# ac_add_options --enable-debug-symbols
# ac_add_options --disable-export-js
# ac_add_options --disable-shared-js
# ac_add_options --disable-intl-api
#
# Didn't add:
# ac_add_options --enable-application=positron
# -> because that's the default app in the positron repo.
# mk_add_options MOZ_OBJDIR=@TOPSRCDIR@/obj.debug.noindex
# -> because it shouldn't be needed now that paths are handled automatically.
#
> ./mach build
```

Positron then built successfully, and I could invoke it to run an app.
